### PR TITLE
Add Country as metric

### PR DESCRIPTION
**Pull request by BetaHuhn**
## Description
- Adds country as metric to both api and ui
- Uses the [geoip-country](https://www.npmjs.com/package/geoip-country) module to get the country code from the clients ip. Everything happens locally to respect the users privacy

## Screenshots
ackeeCountryPull
ackeeCountryPull2

### CODE REVIEW

File: src/schemas/Record.js [hunk 1]
```diff
@@ -34,6 +34,10 @@ const schema = new mongoose.Schema({
 		minlength: 2,
 		maxlength: 2
 	},
+	clientCountry: {
+		type: String,
+		required: false
+	},
```
Comment (RomainCscn) [inline src/schemas/Record.js:40]: You don't have to put the required, it is by default false.
```suggestion
	clientCountry: {
		type: String
	},
```

**Comment by electerious** [comment]
This is a great PR and I would really like to merge it, but what bothers me is that [node-geoip-country](https://github.com/sapics/node-geoip-country#built-in-updater) has a build-in updater and you need to ensure that it updates every week. You also need a MaxMind account and API key.

> The solution we have chosen is to introduce a new end-user license agreement containing the relevant data processing provisions that both we, as the business providing data, and you, as the third party user of the data, need to comply with applicable data privacy regulations. Additionally, by requiring a MaxMind account and contact information from you, we will be able to communicate all valid “Do Not Sell” requests to you as we receive them.

See https://blog.maxmind.com/2019/12/18/significant-changes-to-accessing-and-using-geolite2-databases/

Not sure if there's a good way to solve this in an open-source and self-hosted app. Any ideas?

**Comment by BetaHuhn** [comment]
I understand what you mean.
> Not sure if there's a good way to solve this in an open-source and self-hosted app. Any ideas?

The only other solution I found is using [IP2Location](https://github.com/ip2location-nodejs/IP2Location). It uses a [database](https://lite.ip2location.com/database/ip-country) file which you have to download seperately for free without an API Key or Account. So no build-in updater. It uses [this](https://creativecommons.org/licenses/by-sa/4.0/) license. [Ip2location](https://lite.ip2location.com/) updates its database monthly.

That's probably the best way to implement this. What are your thoughts?

**Comment by pedrogpimenta** [comment]
The ip2location solution is the one that looks best for this case, thanks for investigating @BetaHuhn.

It does add a little bit of size to the project, ranging from 1.5 MB (countries only) and 43.77 MB (much more data, probably unneeded). 1.5 is not that much, I think the benefits are very good.

As a small hobby-project builder, location is one of the most exciting thing for me to see, know where users are coming from :)

**Comment by BetaHuhn** [comment]
> As a small hobby-project builder, location is one of the most exciting thing for me to see, know where users are coming from :)

I have been running Ackee with the PR for a few weeks now and can confirm this. Was pretty cool to see that my small projects have users from the other side of the world 😄

**Comment by electerious** [comment]
I'm currently restructuring a lot for the new dashboard. This feature won't make it into a release soon, but I will keep an eye on it. An optional ip2location integration would we good and I guess there're more people looking for such a feature :)

**Comment by stephenafamo** [comment]
Hello, just curious what the status of this is? Are there still plans to implement?

**Comment by electerious** [comment]
There're currently no plans to implement this feature. Mainly because of what I mentioned earlier: https://github.com/electerious/Ackee/pull/105#issuecomment-616098682

It's a great PR, but it adds a complexity I (currently) don't want to add to Ackee.

**Pull request closed without merging**